### PR TITLE
Track whether or not let expressions failed to solve in solver

### PR DESCRIPTION
**Pull request by abadams**
After mutating an expression, the solver needs to know two things:

1) Did the expression contain the variable we're solving for 
2) Was the expression successfully "solved" for the variable. I.e. the variable only appears once in the leftmost position. We need to know this to know property 1 of any subexpressions (e.g. does the right child of the expression contain the variable). This drives what transformations we do in ways that are guaranteed to terminate and not take exponential time.

We were tracking property 1 through lets but not property 2, and this meant we were doing unhelpful transformations in some cases. I found a case in the wild where this made a pipeline take > 1 hour to compile (I killed it after an hour). It may have been in an infinite transformation loop, or it might have just been exponential. Not sure.

As a drive-by while trying to figure out why it was taking so long I also made remainder checking cheaper in the common case of division by a constant, by avoiding a call to can_prove.

### CODE REVIEW

File: src/Solve.cpp [hunk 1]
```diff
@@ -44,18 +44,22 @@ class SolveExpression : public IRMutator {
         map<Expr, CacheEntry, ExprCompare>::iterator iter = cache.find(e);
         if (iter == cache.end()) {
             // Not in the cache, call the base class version.
-            debug(4) << "Mutating " << e << " (" << uses_var << ")\n";
+            debug(4) << "Mutating " << e << " (" << uses_var << ", " << failed << ")\n";
             bool old_uses_var = uses_var;
```
Comment (steven-johnson) [inline src/Solve.cpp:48]: I sure which C++ had a terse way to express this pattern (that wasn't an RIAA class)...

**Comment by steven-johnson** [comment]
Does this need new test(s)?

**Comment by abadams** [comment]
It's a failure to scale gracefully with large inputs. For the correctness of it, the existing solver tests are fine.

**Comment by abadams** [comment]
Looks like there are some real failures to investigate

**Comment by steven-johnson** [comment]
I assume this is still in-progress?

**Comment by abadams** [comment]
Yes

**Comment by abadams** [comment]
The test failure was caused by use of an uninitialized value because I didn't check a return value. I added the must use result macro to reduce_expr_modulo to prevent this problem in future.

File: src/ModulusRemainder.h [hunk 2]
```diff
@@ -83,8 +85,8 @@ ModulusRemainder modulus_remainder(const Expr &e, const Scope<ModulusRemainder>
 /** Reduce an expression modulo some integer. Returns true and assigns
  * to remainder if an answer could be found. */
 ///@{
-bool reduce_expr_modulo(const Expr &e, int64_t modulus, int64_t *remainder);
-bool reduce_expr_modulo(const Expr &e, int64_t modulus, int64_t *remainder, const Scope<ModulusRemainder> &scope);
+HALIDE_MUST_USE_RESULT bool reduce_expr_modulo(const Expr &e, int64_t modulus, int64_t *remainder);
+HALIDE_MUST_USE_RESULT bool reduce_expr_modulo(const Expr &e, int64_t modulus, int64_t *remainder, const Scope<ModulusRemainder> &scope);
```
Comment (steven-johnson) [inline src/ModulusRemainder.h:89]: "must use result" should really be the default in the language, with the current behavior opt-in :-/

**Review by steven-johnson — APPROVED** [review]

**Pull request merged**